packages/apollo/src/utils/lists.ts

```diff
@@ -67,26 +105,58 @@ export class ListHelper {
             this.skipCacheOnRead = options.cacheOptions.skipOnRead
         }
 
-        bindAll(this, '_readPage', '_networkOnlyRead', 'getReadFunction', 'mergeLists')
+        bindAll(this, '_readSlicePage', '_readAllList', '_networkOnlyRead', 'getReadFunction', 'mergeLists')
     }
 
     /**
-     * Read function, which implements paginated reading from cache
+     * Read function that implements paginated reading from cache.
+     * Returns a slice of the cached list based on pagination arguments.
+     * Includes validation of data integrity by checking for broken references
+     * (references to objects that don't exist in the cache).
      */
-    private _readPage<TData, TOptions extends FieldFunctionOptions>(
+    private _readSlicePage<TData, TOptions extends FieldFunctionOptions>(
         existing: ReadonlyArray<TData> | undefined,
         options: TOptions
     ): Array<TData> | undefined {
-        const skip: number = get(options, ['args', this.skipArgName], 0)
-        const first: number = get(options, ['args', this.firstArgName], 0)
+        const skip: number = options?.args?.[this.skipArgName] || 0
+        const first: number = options?.args?.[this.firstArgName] || 0
 
         // NOTE: partial data (skip only) is not enough for case
         // where you fetch (page + 1) item to show "Next page" pagination event
         if (!existing || existing.length < skip + first) {
             return undefined
         }
 
-        return existing.slice(skip, skip + first)
+        const sliceList = existing.slice(skip, skip + first).filter(Boolean)
+
+        const cache: NormalizedCacheObject | undefined = options?.cache?.extract?.()
+
+        if (!cache || hasBrokenLinks(sliceList, cache)) return undefined
+
+        return sliceList
+    }
+
+    /**
+     * Read function that returns the complete list from cache without pagination.
+     * Includes validation of data integrity by checking for broken references
+     * (references to objects that don't exist in the cache).
+     */
+    private _readAllList<TData, TOptions extends FieldFunctionOptions>(
+        existing: ReadonlyArray<TData> | undefined,
+        options: TOptions
+    ): Array<TData> | undefined {
+        if (!existing) {
+            return undefined
+        }
+
+        const cache: NormalizedCacheObject | undefined = options?.cache?.extract?.()
+
+        const list = existing?.filter(Boolean)
```

Comment:
You have already checked above that the list exists

```suggestion
        const list = existing.filter(Boolean)
```